File: app/src/main/java/yoyo/app/android/com/yoyoapp/Trip/booking/BookingFirstFragment.java
package yoyo.app.android.com.yoyoapp.Trip.booking;


import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.hbb20.CountryCodePicker;
import yoyo.app.android.com.yoyoapp.DataModels.Traveller;
import yoyo.app.android.com.yoyoapp.R;
import yoyo.app.android.com.yoyoapp.Trip.adapter.TravellerRecyclerviewAddapter;

public class BookingFirstFragment extends Fragment {

    private static final String TAG = "BookingFirstFragment";
    private RecyclerView recyclerView;
    private TravellerRecyclerviewAddapter travellerRecyclerviewAddapter;
    public  EditText mobilenumberEditText, emailEditText ,fullNameEditText;
    private TextView passengerCount;
    private ImageView minusImageview, plusImageview;
    private CountryCodePicker countryCodePicker;
    private int passengerNum = 1;

    private View view;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_booking_first2, container, false);
        init();
        setupContactViews();
        setupMobileNumber();
        setupRecyclerview();
        plusImageview.setOnClickListener(v -> addNumber());
        minusImageview.setOnClickListener(v -> reduceNumber());


        return view;
    }

    private void setupContactViews() {
        mobilenumberEditText.setText(((BookingActivity)getActivity()).mobileNumberString);
        emailEditText.setText(((BookingActivity)getActivity()).emailString);
        fullNameEditText.setText(((BookingActivity)getActivity()).fullNameString);
        fullNameEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.length() != 0) {
                    ((BookingActivity)getActivity()).fullNameString = s.toString();
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
        emailEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.length() != 0) {
                    ((BookingActivity)getActivity()).emailString = s.toString();
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
        mobilenumberEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.length() != 0) {
                    ((BookingActivity)getActivity()).mobileNumberString = s.toString();
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void reduceNumber() {
        if (passengerNum > 1) {
            passengerNum--;
            ((BookingActivity)getActivity()).passerngerNumLiveData.postValue(passengerNum);
            passengerCount.setText(String.valueOf(passengerNum));
            travellerRecyclerviewAddapter.notifyDataSetChanged();
        }
        if (passengerNum == 1)
        {
            minusImageview.setImageDrawable(getResources().getDrawable(R.drawable.ic_remove_circle_outline_light_24dp));
        }
    }

    private void addNumber() {
        if (passengerNum < 9) {
            passengerNum++;
            ((BookingActivity)getActivity()).passerngerNumLiveData.postValue(passengerNum);
            passengerCount.setText(String.valueOf(passengerNum));
            travellerRecyclerviewAddapter.notifyDataSetChanged();
        }
        if (passengerNum ==2)
        {
            appearView(minusImageview, passengerCount);
        }
    }
    private void init() {
        recyclerView = view.findViewById(R.id.rv_booking);
        mobilenumberEditText = view.findViewById(R.id.et_traveller_details_mobile);
        emailEditText = view.findViewById(R.id.et_traveller_details_email);
        countryCodePicker = view.findViewById(R.id.ccp_traveller_details);
        minusImageview = view.findViewById(R.id.iv_bookingfirst_minus);
        plusImageview = view.findViewById(R.id.iv_bookingfirst_plus);
        passengerCount = view.findViewById(R.id.tv_bookingfirst_num);
        fullNameEditText = view.findViewById(R.id.et_booking_contact_name);
    }

    // get country code for mobile number
    private void setupMobileNumber() {
        ((BookingActivity)getActivity()).mobileNumberCode = countryCodePicker.getDefaultCountryCodeWithPlus();
        if ( ((BookingActivity)getActivity()).countryCode != null) {
            countryCodePicker.setCountryForNameCode(((BookingActivity)getActivity()).countryCode);
        }
        countryCodePicker.setOnCountryChangeListener(new CountryCodePicker.OnCountryChangeListener() {
            @Override
            public void onCountrySelected() {
                String s = ((BookingActivity)getActivity()).countryCode = countryCodePicker.getSelectedCountryNameCode();
                String s1= ((BookingActivity)getActivity()).mobileNumberCode = countryCodePicker.getSelectedCountryCodeWithPlus();
            }
        });
    }

    // setup traveller recycler view
    private void setupRecyclerview() {
        travellerRecyclerviewAddapter = new TravellerRecyclerviewAddapter(((BookingActivity)getActivity()).travellers, getActivity(), new TravellerRecyclerviewAddapter.OnItemSelected() {
            @Override
            public void onSendResult(Traveller traveller, int position) {
                // when the user on one traveller item clicked, call expandbottomsheet method in booking activity
                Log.d(TAG, "onSendResult: " + traveller.getFirstName());
                sendToTravellerInfoPage(position);
            }
        });
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setAdapter(travellerRecyclerviewAddapter);
    }

    private void sendToTravellerInfoPage(int position) {
        Bundle bundle = new Bundle();
        bundle.putInt("position",position);

        TravellerInfoFragment travellerInfoFragment = new TravellerInfoFragment();
        travellerInfoFragment.setArguments(bundle);
        FragmentTransaction fragmentTransaction = getFragmentManager().beginTransaction();
        fragmentTransaction.add(R.id.fl_booking,travellerInfoFragment);
        fragmentTransaction.addToBackStack("travellerInfo");
        fragmentTransaction.commit();
    }

    public void appearView(ImageView imageView, TextView textView)
    {
        imageView.setImageDrawable(getResources().getDrawable(R.drawable.ic_remove_circle_outline_black_24dp));
        textView.setTextColor(getResources().getColor(R.color.colorPrimary));
    }
}

File: app/src/main/java/yoyo/app/android/com/yoyoapp/trip/booking/firstPage/BookingFirstFragment.kt
package yoyo.app.android.com.yoyoapp.trip.booking.firstPage

import android.os.Bundle
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.activityViewModels
import androidx.fragment.app.viewModels
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import es.dmoral.toasty.Toasty
import kotlinx.android.synthetic.main.fragment_booking_first2.view.*
import kotlinx.android.synthetic.main.fragment_booking_first2.view.iv_sign_out_back
import kotlinx.android.synthetic.main.fragment_booking_first2.view.progressbar_booking_book
import yoyo.app.android.com.yoyoapp.DataModels.Traveller
import yoyo.app.android.com.yoyoapp.MainActivity
import yoyo.app.android.com.yoyoapp.R
import yoyo.app.android.com.yoyoapp.SharedDataViewModel
import yoyo.app.android.com.yoyoapp.trip.adapter.TravellerRecyclerviewAddapter
import yoyo.app.android.com.yoyoapp.trip.api.BookScheduleRequest
import yoyo.app.android.com.yoyoapp.trip.booking.BookingSecondFragment
import yoyo.app.android.com.yoyoapp.trip.booking.TravellerInfoFragment
import yoyo.app.android.com.yoyoapp.trip.dialog.ExitDialogFragment
import java.util.ArrayList

class BookingFirstFragment : Fragment() {

    var travellers: ArrayList<Traveller> = ArrayList()
    private var travellerRecyclerAdapter: TravellerRecyclerviewAddapter? = null
    private var passengerNum = 2
    private var minCapacity = 1
    private val bookingViewModel by viewModels<BookingViewModel>()
    private val sharedDataViewModel by activityViewModels<SharedDataViewModel>()
    private var price: Double = 0.0
    private lateinit var scheduleId: String
    private var mobileNumberCode: String? = null
    private lateinit var bookScheduleRequest: BookScheduleRequest

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val res = inflater.inflate(R.layout.fragment_booking_first2, container, false)

        init(res)
        observeTravellers()
        setupTravellersArray()
        res.iv_bookingfirst_plus.setOnClickListener { addTravellersNumber()}
        res.iv_bookingfirst_minus.setOnClickListener { reduceTravellersNumber()}
        res.button_booking_continue.setOnClickListener{ setupContinueButton()}
        res.iv_sign_out_back.setOnClickListener{ setupExit()}
        setupRecycler(res)
        setupMobileNumber(res)

        return res
    }

    private fun observeTravellers() {
        sharedDataViewModel.travellersList.observe(context as LifecycleOwner, Observer {
            travellers = it
            travellerRecyclerAdapter?.notifyDataSetChanged()
        })
    }

    private fun setupTravellersArray() {
        for (i in 0 until minCapacity) {
            val traveller = Traveller()
            sharedDataViewModel.addTraveller(traveller)
        }
    }

    private fun setupContinueButton() {
        val fieldsAreNotNull = checkingEmptyItems(view?.et_booking_contact_name?.text.toString(),
            view?.et_traveller_details_email?.text.toString(),
            view?.et_traveller_details_mobile?.text.toString(), travellers)
        if (fieldsAreNotNull) {
            bookFlight()
        }
    }

    private fun bookFlight() {
        view?.progressbar_booking_book?.visibility = View.VISIBLE
        setupBookingRequestObject()
        bookingViewModel.initBookingRequest(scheduleId, bookScheduleRequest)
        bookingViewModel.paymentUrlMutableLiveData.observe(activity!!, Observer { newurl ->
            if (newurl != null) {
                val bundle = Bundle()
                bundle.putString("name", view?.et_booking_contact_name?.text.toString())
                bundle.putInt("travellerNumber", travellers.size)
                bundle.putDouble("price", price)
                bundle.putString("url", newurl)

                val fragmentTransaction = fragmentManager?.beginTransaction()
                val bookingSecondFragment = BookingSecondFragment()
                bookingSecondFragment.arguments = bundle
                fragmentTransaction?.replace(R.id.main_container, bookingSecondFragment)
                fragmentTransaction?.commit()
            }
            view?.progressbar_booking_book?.visibility = View.GONE
        })
    }

    private fun setupBookingRequestObject() {
        val leaderTraveller = BookScheduleRequest.LeaderTraveller(
            view?.et_traveller_details_email?.text.toString()
            , view?.et_booking_contact_name?.text.toString()
            , mobileNumberCode + view?.et_traveller_details_mobile?.text.toString())

        val travellerCompanions = ArrayList<BookScheduleRequest.CompanionTraveller>()
        for (traveller in travellers) {
            traveller.apply {
                travellerCompanions.add(
                    BookScheduleRequest.CompanionTraveller(
                        dateOfBirthTimeStamp.toInt(),
                        firstName,
                        gender,
                        lastName,
                        nationalityCode,
                        nationality,
                        passportNumber
                    )
                )
            }
        }
        bookScheduleRequest = BookScheduleRequest(travellerCompanions,leaderTraveller)
    }

    private fun checkingEmptyItems(
        fullNameString: String?,
        emailString: String?,
        phoneNumberString: String?,
        travellers: ArrayList<Traveller>
    ): Boolean {
        if (fullNameString.isNullOrEmpty()) {
            Toasty.error(activity!!, "full name can not be empty").show()
            return false
        }
        if (emailString.isNullOrEmpty()) {
            Toasty.error(activity!!, resources.getString(R.string.email_cannot_be_empty)).show()
            return false
        }
        if (phoneNumberString.isNullOrEmpty()) {
            Toasty.error(activity!!, resources.getString(R.string.phone_number_cannot_be_empty)).show()
            return false
        }
        for (i in travellers.indices) {
            if (travellers[i].firstName == null) {
                Toasty.error(activity!!, "Please complete all traveller's data").show()
                return false
            }
        }
        if (! view?.et_traveller_details_email?.text.toString()
                .matches("^[A-Za-z0-9_.]+[@][A-Za-z.]+$".toRegex())) {
            Toasty.error(activity!!, "Your Email Address is incorrect").show()
            return false
        }
        return true
    }


    private fun reduceTravellersNumber() {
        if (passengerNum > minCapacity) {
            passengerNum --
            travellers.removeAt(travellers.size - 1)
            view?.tv_bookingfirst_num?.text = passengerNum.toString()
            travellerRecyclerAdapter?.notifyDataSetChanged()
        }
        if (passengerNum == minCapacity) {
            view?.iv_bookingfirst_minus?.setImageDrawable(resources.getDrawable(R.drawable.ic_remove_circle_outline_light_24dp))
        }
    }

    private fun addTravellersNumber() {
        if (passengerNum < 9) {
            passengerNum++
            val traveller = Traveller()
            travellers.add(traveller)
            view?.tv_bookingfirst_num?.text = passengerNum.toString()
            travellerRecyclerAdapter?.notifyDataSetChanged()
        }
        if (passengerNum >= minCapacity + 1) {
            appearMinusView(view?.iv_bookingfirst_minus, view?.tv_bookingfirst_num)
        }
    }

    private fun init(res: View) {
        minCapacity = arguments?.getInt("minCapacity") ?: 1
        price = arguments?.getDouble("price") ?: 0.0
        scheduleId = arguments?.getString("scheduleId") ?: ""
        passengerNum = travellers.size
        res.tv_bookingfirst_num.text = passengerNum.toString()
    }

    // get country code for mobile number
    private fun setupMobileNumber(res: View) {
        mobileNumberCode = res.ccp_traveller_details.defaultCountryCodeWithPlus
        res.ccp_traveller_details.setOnCountryChangeListener {
            mobileNumberCode = res.ccp_traveller_details.selectedCountryCodeWithPlus
        }
    }

    private fun setupRecycler(res: View) {
        travellerRecyclerAdapter = TravellerRecyclerviewAddapter(travellers, activity,
            TravellerRecyclerviewAddapter.OnItemSelected { traveller, position ->
                sendToTravellerInfoPage(position)
            })
        val linearLayoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        res.rv_booking.layoutManager = linearLayoutManager
        res.rv_booking.adapter = travellerRecyclerAdapter
    }

    private fun sendToTravellerInfoPage(position: Int) {
        val bundle = Bundle()
        bundle.putInt("position", position)

        val travellerInfoFragment = TravellerInfoFragment()
        travellerInfoFragment.arguments = bundle
        (context as MainActivity).showFragment(this, travellerInfoFragment, "traveller_Info", false)

    }

    private fun appearMinusView(imageView: ImageView?, textView: TextView?) {
        imageView?.setImageDrawable(resources.getDrawable(R.drawable.ic_remove_circle_outline_black_24dp))
        textView?.setTextColor(resources.getColor(R.color.colorPrimary))
    }

    private fun setupExit() {
        val exitDialogFragment = ExitDialogFragment()
        fragmentManager?.let { exitDialogFragment.show(it, "exit") }
    }

    override fun onPause() {
        sharedDataViewModel.resetBookingTravellers()
        super.onPause()
    }
}
